pkg/utils/requiredfieldsvalidator/validator.go

```diff
@@ -0,0 +1,141 @@
+package requiredfieldsvalidator
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/instaclustr/operator/pkg/models"
+)
+
+func ValidateRequiredFields(v any) error {
+	val := reflect.ValueOf(v)
+
+	if val.Kind() == reflect.Ptr {
+		val = val.Elem()
+	}
+
+	if val.Kind() != reflect.Struct {
+		return fmt.Errorf("expected struct type - got the %s", val.Kind())
+	}
+
+	err := validateValue(val)
+	if err != nil {
+		return fmt.Errorf("error have occured: %w", err)
+	}
+
+	return nil
+}
+
+func validateValue(val reflect.Value) error {
+	kind := val.Kind()
+	_ = kind
+	switch val.Kind() {
+	case reflect.Struct:
+		return validateStruct(val)
+
+	case reflect.Array, reflect.Slice:
+		return validateArrayOrSlice(val)
+
+	case reflect.Map:
+		return validateMap(val)
+
+	case reflect.Ptr:
+		return validatePointer(val)
+
+	case reflect.Bool:
+		return nil
+
+	default:
+		if val.IsZero() {
+			return fmt.Errorf("required field is empty")
+		}
+	}
+	return nil
+}
+
+func isStructPtr(fieldValue reflect.Value) bool {
+	return fieldValue.Kind() == reflect.Ptr && !fieldValue.IsNil() && fieldValue.Elem().Kind() == reflect.Struct
+}
+
+func tagContainsOmitempty(tag string) bool {
+	return strings.Contains(tag, models.OmitemptyTag)
+}
+
+func validateStruct(val reflect.Value) error {
+	for i := 0; i < val.NumField(); i++ {
+		field := val.Type().Field(i)
+		fieldValue := val.Field(i)
+
+		if (fieldValue.Kind() == reflect.Array || fieldValue.Kind() == reflect.Slice) && fieldValue.Len() > 0 {
+			err := validateValue(fieldValue)
+			if err != nil {
+				return fmt.Errorf("%s, %w", field.Name, err)
+			}
+		}
+
+		if tag, ok := field.Tag.Lookup(models.JsonTag); ok && tagContainsOmitempty(tag) {
+			if fieldValue.Kind() == reflect.Ptr || fieldValue.Kind() == reflect.Map {
+				if !fieldValue.IsZero() {
+					err := validateValue(fieldValue)
+					if err != nil {
+						return fmt.Errorf("%s, %w", field.Name, err)
+					}
+				}
+			}
+			continue
+		}
+
+		err := validateValue(fieldValue)
+		if err != nil {
+			return fmt.Errorf("%s, %w", field.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func validateArrayOrSlice(val reflect.Value) error {
+	if val.Len() < 1 {
+		return fmt.Errorf("len is 0")
+	}
+	for i := 0; i < val.Len(); i++ {
+		fieldValue := val.Index(i)
+		if fieldValue.Kind() == reflect.Struct ||
+			isStructPtr(fieldValue) {
+			err := validateValue(val.Index(i))
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+
+	return nil
+}
+
+func validateMap(val reflect.Value) error {
+	if val.Len() < 1 {
+		return fmt.Errorf("empty map")
+	}
+	iter := val.MapRange()
+	for iter.Next() {
+		v := iter.Value()
+		if v.Kind() == reflect.Struct ||
+			isStructPtr(v) {
+			err := validateValue(v)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func validatePointer(val reflect.Value) error {
+	if val.IsNil() {
+		return fmt.Errorf("pointer to nil")
+	}
+	return validateValue(val.Elem())
+}
```

Comment:
Your code is good, I just have a suggestion, I didn't test it but you can see on my basic code structure and tell me WDYT. I think it would just a bit simplier

```suggestion

func validateStruct(val reflect.Value) error {
    for i := 0; i < val.NumField(); i++ {
        field := val.Type().Field(i)
        fieldValue := val.Field(i)

        if tag, ok := field.Tag.Lookup(JsonTag); ok && strings.Contains(tag, OmitemptyTag) && fieldValue.IsZero() {
            continue
        }

        if err := validateField(fieldValue); err != nil {
            return fmt.Errorf("%s: %w", field.Name, err)
        }
    }
    return nil
}

func validateField(fieldValue reflect.Value) error {
    switch fieldValue.Kind() {
    case reflect.Struct:
        return validateStruct(fieldValue)
    case reflect.Array, reflect.Slice:
        if fieldValue.Len() == 0 {
            return fmt.Errorf("array or slice is empty")
        }
        for i := 0; i < fieldValue.Len(); i++ {
            if err := validateField(fieldValue.Index(i)); err != nil {
                return err
            }
        }
    case reflect.Map:
        if fieldValue.Len() == 0 {
            return fmt.Errorf("map is empty")
        }
        for _, key := range fieldValue.MapKeys() {
            if err := validateField(fieldValue.MapIndex(key)); err != nil {
                return err
            }
        }
    case reflect.Ptr, reflect.Interface:
        if fieldValue.IsNil() {
            return fmt.Errorf("nil pointer or interface")
        }
        return validateField(fieldValue.Elem())
    default:
        if fieldValue.IsZero() {
            return fmt.Errorf("field is empty")
        }
    }
    return nil
}
```

Reply:
![image](https://github.com/instaclustr/operator/assets/83217719/f84afd49-e975-4cd6-86ff-3406a74fbc0a)
![image](https://github.com/instaclustr/operator/assets/83217719/0c7b0b73-57ae-4043-ad9e-dec6a7d239ab)
it hasn't pass the tests on arrays that initialized but empty